cranelift: Use `cranelift-jit` in runtests

afonso360 opened the pull request:
👋  Hey,

This PR changes our run tests to run via `cranelift-jit`. Its based on top of #4450 so its probably not worth doing any serious review until that is merged.

Using `cranelift-jit` in run tests allows us to preform relocations and
libcalls. This is important since some instruction lowerings fallback
to libcall's when an extension is missing, or when it's too complicated
to implement manually.

This is also a first step to being able to test `call`'s between functions
in the runtest suite. It should also make it easier to eventually test
TLS relocations, symbol resolution and ABI's.

Another benefit of this is that we also get to test the JIT more, since
it now runs the runtests, and gets some fuzzing via `fuzzgen` (which
uses the `SingleFunctionCompiler`).

This change causes regressions in terms of runtime for the filetests.
I haven't done any serious benchmarking but what I've been seeing is
that it now takes about ~3 seconds to run the testsuite while it
previously took around 2 seconds.

Edit: 
This also closes #3030 since we can now do relocations in runtests

afonso360 commented on cranelift/jit/src/backend.rs:
```diff
@@ -893,7 +893,7 @@ fn lookup_with_dlsym(name: &str) -> Option<*const u8> {
     use windows_sys::Win32::System::LibraryLoader;
 
-    const MSVCRT_DLL: &[u8] = b"msvcrt.dll\0";
+    const UCRTBASE: &[u8] = b"ucrtbase.dll\0";
```
I'm slightly concerned about this change.

`msvcrt` does not define `ceilf`/`floorf`/`nearbyintf`/`truncf`, which are some of the LibCall's that we define. I cannot find the dll that defines those functions pre VS2015.

After VS2015 Microsoft created the Universal C Runtime which defines `ucrtbase.dll` that does contain all of those functions. However that only became standard with Windows 10 in 2015. Previous versions of Windows can have that dll, but only if they install Visual Studio 2015.

We cannot search `msvcrt` and fallback to `ucrtbase` since mixing the two runtimes will cause weird behaviour, probably at the worst time possible.

Does anyone know where we can find those functions? Or alternativley what are the minimum requirements that we have for cranelift and can we upgrade to UCRT?

jameysharp replied:
I'm told Windows 8.1 reaches end of life in January, which seems to me like a good enough reason for Cranelift to require Windows 10. But others may disagree.

Since we're looking at using the libm crate for #4517, I thought I'd check if we could get these functions from there, instead of worrying about Windows C runtime versions. It does have `ceilf`, `floorf`, and `truncf`, but unfortunately not `nearbyintf`, so I'm not sure if that helps with this question.

afonso360 replied:
@cfallin I would like to move forward with this PR since it's blocking some stuff on the fuzzer (getting function calls working).

Would @jameysharp 's reasoning that pretty soon all supported windows versions are going to have `ucrt` be sufficient? An alternative could be to keep using `msvcrt`, but we would need to disable some ops on the runtest suite, which is also not ideal.

We can also go with polyfill's as @jameysharp mentioned, I don't like this approach since we would probably keep needing to patch holes in msvcrt. In https://github.com/bjorn3/rustc_codegen_cranelift/pull/1254 we also found that `fabsf` is missing from msvcrt, so that's going to be another one that we need. With the bonus that we know that one is [definitely broken in libm](https://github.com/bytecodealliance/wasmtime/pull/4517#issuecomment-1195138197).

cfallin replied:
Yeah, I think that's fine; Win10 as a minimum requirement here is probably an acceptable compromise, given the context.

cfallin reviewed (commented):
Thanks for this! The change overall looks reasonable. I'm not sure about the Windows runtime issue; tagged @peterhuene for more thoughts.

afonso360 commented:
With the CRT issue resolved I think this is ready to merge.

The only changes since the last review were a rebase and enabling the FMA tests that we couldn't enable in #4460.

Thanks for reviewing!

jameysharp commented on cranelift/filetests/src/function_runner.rs:
```diff
@@ -66,27 +66,47 @@ impl SingleFunctionCompiler {
     ///  - compile a `Trampoline` for the [Function]'s signature (or used a cached `Trampoline`;
     ///    this makes it possible to call functions when the signature is not known until runtime.
-    pub fn compile(&mut self, function: Function) -> Result<CompiledFunction, CompilationError> {
+    pub fn compile(self, function: Function) -> Result<CompiledFunction, CompilationError> {
```
Changing `SingleFunctionCompiler::compile` from taking `&mut self` to just `self` feels unfortunate, since it forces callers to build a new `TargetIsa` for every function they compile. What would you think of changing `JITBuilder` to take a shared reference instead? Either `&dyn TargetIsa` or `Rc<dyn TargetIsa>`, whichever is easier. In the latter case, `SingleFunctionCompiler` would also take an `Rc` instead of a `Box`.

afonso360 replied:
I remember (sorry, this was a while ago) trying to change to `&dyn TargetIsa`, but it brought a bunch of changes due to lifetimes, and I gave up on that path.

I didn't  try `Rc<>` but see the comment down below.

jameysharp commented on cranelift/fuzzgen/src/function_generator.rs:
```diff
@@ -691,7 +691,7 @@ where
 
         let mut fn_builder_ctx = FunctionBuilderContext::new();
-        let mut func = Function::with_name_signature(ExternalName::user(0, 0), sig.clone());
+        let mut func = Function::with_name_signature(ExternalName::user(0, 1), sig.clone());
```
What is this change for?

afonso360 replied:
The name `user(0, 0)` was the trampoline IIRC and the JIT was complaining about it. I'm not too sure, but I can go check.

jameysharp commented on cranelift/filetests/src/test_run.rs:
```diff
@@ -118,20 +132,20 @@ impl SubTest for TestRun {
                 trace!("Parsed run command: {}", command);
 
-                let compiled_fn = compiler.compile(func.clone().into_owned())?;
+                // We can't use the requested ISA directly since it does not contain info
+                // about the operating system / calling convention / etc..
+                //
+                // Copy the requested ISA flags into the host ISA and use that.
+                let isa = build_host_isa(false, context.flags.clone(), requested_isa.isa_flags());
+
+                let compiled_fn =
+                    SingleFunctionCompiler::new(isa).compile(func.clone().into_owned())?;
```
I'd prefer not to move this code into the loop, if my earlier suggestion on shared references to the `TargetIsa` works out.

jameysharp commented on cranelift/src/run.rs:
```diff
@@ -85,12 +85,11 @@ fn run_file_contents(file_contents: String) -> Result<()> {
         for comment in comments {
             if let Some(command) = parse_run_command(comment.text, &func.signature)? {
-                let compiled_fn = compiler.compile(func.clone())?;
+                let isa = create_target_isa(&test_file.isa_spec)?;
+                let compiled_fn = SingleFunctionCompiler::new(isa).compile(func.clone())?;
```
I'd prefer not to move this code into the loop, if my earlier suggestion on shared references to the `TargetIsa` works out.

jameysharp commented on cranelift/filetests/src/function_runner.rs:
```diff
@@ -66,27 +66,47 @@ impl SingleFunctionCompiler {
+        let name = format!("{}", function.name);
+        let func_id = module.declare_function(&name, Linkage::Local, &function.signature)?;
+
+        // Build and declare the trampoline in the module
+        let trampoline_name = format!("{}", trampoline.name);
+        let trampoline_id =
+            module.declare_function(&trampoline_name, Linkage::Local, &trampoline.signature)?;
```
If it works out to keep the `SingleFunctionCompiler` reusable, do you think it'd be worth keeping the cache of trampolines around too?

jameysharp reviewed (approved):
Overall I think this is great. I'd just like to take a little more time on ownership of `TargetIsa` and reuse of `SingleFunctionCompiler`. I'm open to merging it as-is if you feel that it's better this way, @afonso360, but it seems to me that it's useful to be able to reuse these.

afonso360 commented:
@jameysharp I agree with your comments, however this is sort of an intermediate PR with the follow up cleaning pretty much all of it.

See #4667, we do away with SingleFunctionCompiler and link the entire test file at once, so the repeated compilation issue goes away.

jameysharp commented:
I hadn't looked at 4667 yet, and that resolves my concerns. Thanks!

uweigand commented:
As of this commit, I'm now seeing test failures on s390x (Ubuntu 20.04), specifically when running `cargo test --doc` in the filetests directory.  The symptom is:
```
Test executable failed (exit status: 101).
stderr:
thread 'main' panicked at 'unable to make memory readable+executable: SystemCall(Os { code: 13, kind: PermissionDenied, message: "Permission denied" })', cranelift/jit/src/memory.rs:178:30
```
when running the sample code in `src/function_runner.rs`.   A backtrace shows that the failing call is:
```
#0  0x000003fffdc8cce0 in mprotect () from /lib64/libc.so.6
#1  0x000002aa0008ee96 in region::os::unix::set_protection (base=0x2aa00847000, size=4096, protection=...) at src/os/unix.rs:30
#2  0x000002aa0008e880 in region::protect::protect (address=0x2aa00847000, size=4096, protection=...) at src/protect.rs:48
#3  0x000002aa00072f6e in cranelift_jit::memory::{impl#2}::set_readable_and_executable::{closure#0} (ptr=0x2aa00847000, len=4096) at cranelift/jit/src/memory.rs:211
#4  0x000002aa00086824 in cranelift_jit::memory::Memory::set_readable_and_executable (self=0x3ffffffd2d0) at cranelift/jit/src/memory.rs:221
#5  0x000002aa00068332 in cranelift_jit::backend::JITModule::finalize_definitions (self=0x3ffffffd260) at cranelift/jit/src/backend.rs:461
#6  0x000002aa00045ae0 in cranelift_filetests::function_runner::TestFileCompiler::compile (self=<error reading variable: value has been optimized out>)
    at cranelift/filetests/src/function_runner.rs:247
```
Note that this attempts to use `mprotect` to grant execute permission to memory allocated on the *heap*.   This is not supported everywhere, and will in particular be prevented by various security hardening settings on Linux.

Instead, you should use separate `mmap` allocations if you need to grant execute permissions.

afonso360 commented:
👋 Hey,

Do the regular filetests pass for you on `main`? We still use `mprotect`, so I'm surprised as to how we haven't detected this sooner.

`cranelift-jit` has the `selinux-fix` feature that switches to `mmap` based allocations, if you try to use that, does that make it pass?

If so, that might be related to: https://github.com/bytecodealliance/wasmtime/issues/4986

uweigand commented:
No, only the doc tests are indeed failing.  I've been debugging this right now, and the difference is that doc test sample program is single-threaded and uses the main heap, where `mprotect` `PROT_EXEC` fails.  However, when running the main filetests, the `clif-util` program is multi-threaded and glibc uses per-thread heaps, which are separately created via `mmap`, and therefore the `mprotect` happens to work ...

How would I use the `selinux-fix` feature with a doc test?

afonso360 commented:
> No, only the doc tests are indeed failing. I've been debugging this right now, and the difference is that doc test sample program is single-threaded and uses the main heap, where mprotect PROT_EXEC fails. However, when running the main filetests, the clif-util program is multi-threaded and glibc uses per-thread heaps, which are separately created via mmap, and therefore the mprotect happens to work ...

That's interesting, and weird!

> How would I use the selinux-fix feature with a doc test?

[This](https://github.com/afonso360/wasmtime/commit/9d72edd9ab411669fa43268e1de70d88c985e8b7) should do it.

uweigand commented:
> > How would I use the selinux-fix feature with a doc test?
> 
> [This](https://github.com/afonso360/wasmtime/commit/9d72edd9ab411669fa43268e1de70d88c985e8b7) should do it.

This fixes the problem for me.  Submitted as https://github.com/bytecodealliance/wasmtime/pull/5204
 
> However I think that we probably should switch to a mmap allocator since that apparently can solve a bunch of other issues with cranelift-jit.

Agreed!
